torch/fx/experimental/proxy_tensor.py

```diff
@@ -1224,32 +1245,62 @@ def _init_modes_from_inputs(self, f, args):
-                self.tracing_mode,
-                pre_dispatch=self.pre_dispatch,
-                _allow_fake_constant=self._allow_fake_constant,
-                _error_on_data_dependent_ops=self._error_on_data_dependent_ops
-            )
+            self._init_modes_from_fx_tracer(self.fx_tracer)
+            yield
+        finally:
+            self._restore_modes(*prev_modes)
+
+    def _init_modes_from_fx_tracer(self, fx_tracer):
+        self.proxy_mode = ProxyTorchDispatchMode(
+            fx_tracer,
+            self.tracing_mode,
+            pre_dispatch=self.pre_dispatch,
+            _allow_fake_constant=self._allow_fake_constant,
+            _error_on_data_dependent_ops=self._error_on_data_dependent_ops
+        )
 
-            if self.pre_dispatch:
-                self.proxy_function_mode = PreDispatchTorchFunctionMode(self.fx_tracer)
+        if self.pre_dispatch:
+            self.proxy_function_mode = PreDispatchTorchFunctionMode(fx_tracer)
 
-            # pre-autograd tracing uses per-dispatch-key modes,
-            # which requires the python dispatcher
-            if self.tracing_mode == "symbolic" or self.pre_dispatch:
-                self.python_dispatcher_mode = enable_python_dispatcher()
+        # pre-autograd tracing uses per-dispatch-key modes,
+        # which requires the python dispatcher
+        if self.tracing_mode == "symbolic" or self.pre_dispatch:
+            self.python_dispatcher_mode = enable_python_dispatcher()
 
-            self.torch_fn_metadata_mode = TorchFunctionMetadataMode(self.fx_tracer)
+        self.torch_fn_metadata_mode = TorchFunctionMetadataMode(fx_tracer)
 
+    @contextmanager
+    def _init_modes_from_parent(self, parent_tracer):
+        # By default, subtracer creates new modes based on parent tracer's config.
+        # However, there are cases where we want to share the same modes with parent tracer
+        # 1. fake_tensor_mode: we want the example value's fake_mode of parent graph and subgraphs to be the same.
+        # 2. tracer.scope_root, proxy_paths, proxy_modules: these are used to construct 'nn_module_stack'
+        #   which has a special logic of handling module aliasing so parent and subgraphs should share the same states.
+        #   For detail, see comments in _ModuleSstackTracer.
+        prev_modes = self._checkpoint_modes()
+        try:
+            self.fake_tensor_mode = parent_tracer.fake_tensor_mode
+
+            def _create_sub_fx_tracer(parent_tracer):
+                if type(parent_tracer) == PythonKeyTracer:
+                    sub_tracer = PythonKeyTracer()
+                elif type(parent_tracer) == _ModuleStackTracer:
+                    sub_tracer = _ModuleStackTracer(parent_tracer.scope_root)
+                    # We want sub_tracer to share the same states as parent tracer
+                    # for below states to correctly compute some metadata
+                    sub_tracer.scope_root = parent_tracer.scope_root
```

Comment:
Add a test for `nn_module_stack`

Reply:
Its a bit difficult to test in this diff: control flow operators turn on torch.compile by default so at the time we're doing reenter_make_fx in Proxy mode, the subgraphs have been flattened. and we lose track of these metadata anyway. This nees a bit more design of keeping original user's metadata when we turn on torch.compile by default.